api/axis.py

```diff
@@ -199,45 +199,61 @@ def update_licensepools_for_identifiers(self, identifiers):
         The book's LicensePool will be updated with current
         circulation information.
         """
-        identifier_strings = self.create_identifier_strings(identifiers)
-        response = self.availability(title_ids=identifier_strings)
         collection = self.collection
-        parser = BibliographicParser(collection)
         remainder = set(identifiers)
-        for bibliographic, availability in parser.process_all(response.content):
+        for bibliographic, availability in self._fetch_remote_availability(identifiers):
             identifier, is_new = bibliographic.primary_identifier.load(self._db)
             if identifier in remainder:
                 remainder.remove(identifier)
-            pool, is_new = availability.license_pool(self._db, collection)
+            pool, is_new = availability.license_pool(self._db, self.collection)
             availability.apply(self._db, pool.collection)
 
         # We asked Axis about n books. It sent us n-k responses. Those
         # k books are the identifiers in `remainder`. These books have
         # been removed from the collection without us being notified.
         for removed_identifier in remainder:
-            pool = identifier.licensed_through_collection(self.collection)
-            if not pool:
-                self.log.warn(
-                    "Was about to reap %r but no local license pool in this collection.",
-                    removed_identifier
-                )
-                continue
-            if pool.licenses_owned == 0:
-                # Already reaped.
-                continue
-            self.log.info(
-                "Reaping %r", removed_identifier
-            )
+            self._reap(removed_identifier)
+
+    def _fetch_remote_availability(self, identifiers):
+        """Retrieve availability information for the specified identifiers.
 
-            availability = CirculationData(
-                data_source=pool.data_source,
-                primary_identifier=removed_identifier,
-                licenses_owned=0,
-                licenses_available=0,
-                licenses_reserved=0,
-                patrons_in_hold_queue=0,
+        :yield: A stream of (Metadata, CirculationData) 2-tuples.
+        """
+        identifier_strings = self.create_identifier_strings(identifiers)
+        response = self.availability(title_ids=identifier_strings)
+        parser = BibliographicParser(self.collection)
+        return parser.process_all(response.content)
+
+    def _reap(self, identifier):
+        """Update our local circulation information to reflect the fact that
+        the identified book has been removed from the remote
+        collection.
+        """
+        collection = self.collection
```

Comment:
I think this is a redundant variable assignment left from earlier, but the variable is then used later in the method.

Reply:
Yeah, I'll leave this one in because `collection` is a property that hits the database.